### ISSUE TITLE
Stats in API

### ISSUE THREAD

**Issue by kantoniak** [pooh-labs/eaterslab#141]

**Comment by kantoniak** [comment]
I had a few thought while working on admin statistics. @Rhantolq @avan1235 please say what do you think about it.

### Occupancy
Cafeteria occupancy cannot be aggregated, because it's value is continuous. API can create a time point for each `CameraEvent` and count events up to this point. For example such call:
```
//api/beta/cafeterias/<id>/occupancy/?from=<from>&to=<to>
<id> = 1
<from> = 2020-05-20T09:37:38.593Z
<to> = 2020-05-20T09:39:38.593Z
```
would return an array of entries:
```
[
  {
    timestamp: '2020-05-20T09:37:38.593Z', // ISO8601 date
    value: {
      relative: 0.86, // 86% occupied
      total: 43       // 43 people inside
    }
  },
  {
    timestamp: '2020-05-20T09:38:38.593Z',
    value: {
      relative: 0.88,
      total: 44
    }
  },
  {
    timestamp: '2020-05-20T09:39:38.593Z',
    value: {
      relative: 0.86,
      total: 43
    }
  },
]
```
This output would mean that:
* X+43 people entered and X people left cafeteria 1 until `2020-05-20T09:37:38.593Z` (inclusive).
* There was an enter/leave event with timestamp `2020-05-20T09:37:38.593Z`
* Somebody then entered cafeteria at '2020-05-20T09:38:38.593Z'
* Somebody then entered cafeteria at '2020-05-20T09:39:38.593Z'

To handle custom occupancy overrides from the admin panel we can have a new type of camera event, let's say `OCCUPANCY_OVERRIDE` with a value which is number set by canteen owner. (This requires slight denormalization, CameraEvent becomes Event with fields `timestamp`, `event_type`, `value` (nullable), `camera_id` (nullable), `cafeteria_id`).

Then algorithm for computing response above would be:
1. For cafeteria `<id>`, fetch all events since last `OCCUPANCY_OVERRIDE` or from the beginning.
2. For each enter/leave event:
2.1 add/subtract from counter, compute relative occupancy;
2.2 add entry to JSON.
3. Remove entries outside `[from, to]` range

If person is not an owner, API call would return only relative occupancy, without `value.total` field.

This can be extended with some batching, making a timestamp every `n` seconds or so:
```
[
  {
    timestamp: '2020-05-20T09:37:38.593Z',
    value: {
      relative: 0.86, // 86% occupied
      total: 43       // 43 people inside
    }
  },
  {
    // Batches X+6 enter events, X leave events
    timestamp: '2020-05-20T09:38:38.593Z',
    value: {
      relative: 0.98,
      total: 49
    }
  },
  {
    // Batches Y enter events, Y+4 leave events
    timestamp: '2020-05-20T09:39:38.593Z',
    value: {
      relative: 0.90,
      total: 45
    }
  },
]
```

**Comment by kantoniak** [comment]
### Average dish review
This data has to be aggregated, but computations are quite easy:
1. Fetch all reviews for dishes from cafeteria `<id>` between `to` and `from`
2. Group by year/month/7 days/day/hour as in docs
3. Get start timestamp for each group (e.g. `2020-05-20T00:00:00.000Z" for a day)
4. Compute average for each group
5. List groups as a response

Sample response for
```
//api/beta/cafeterias/<id>/avg_dish_review/?from=<from>&to=<to>&group_by=<group_by>
<id> = 1
<from> = 2020-05-18T00:00:00.000Z
<to> = 2020-05-21T00:00:00.000Z
<group_by> = day
```
```
[
  {
    timestamp: '2020-05-18T00:00:00.000Z',
    value: 3.84
  },
  {
    timestamp: '2020-05-19T00:00:00.000Z',
    value: 4.7
  },
  {
    timestamp: '2020-05-20T00:00:00.000Z',
    value: 4.32
  },
]
```

**Comment by kantoniak** [comment]
We will need some kind of grouping in occupancy. Image below shows two graphs with entries each hour in the past year. Website takes 260 MB of browser memory vs 50 MB when data comes only from past week. Also, transferring such amounts of data would take very long.

![stats-performance](https://user-images.githubusercontent.com/17510686/83000197-538ba800-a00a-11ea-9615-4fae15b05b7a.png)

**Comment by avan1235** [comment]
Working on it on [avan1235/stats-api](https://github.com/pooh-labs/eaterslab/tree/avan1235/stats-api) branch. Let's see the current progress and test with inserting new camera events form device

**Comment by kantoniak** [comment]
For the record, @avan1235 implemented occupancy in #155. @Rhantolq, we still need average review in stats.
